SU2_CFD/src/drivers/CDiscAdjSinglezoneDriver.cpp

```diff
@@ -431,31 +586,255 @@ void CDiscAdjSinglezoneDriver::SecondaryRecording(){
+
+      /*--- Surface based obj. function ---*/
+      direct_output->SetHistoryOutput(geometry, solver, config, config->GetTimeIter(),
+                                      config->GetOuterIter(), config->GetInnerIter());
+      ObjFunc += solver[FLOW_SOL]->GetTotal_ComboObj();
+      break;
+
+    case MAIN_SOLVER::DISC_ADJ_HEAT:
+      direct_output->SetHistoryOutput(geometry, solver, config, config->GetTimeIter(),
+                                      config->GetOuterIter(), config->GetInnerIter());
+      ObjFunc = solver[HEAT_SOL]->GetTotal_ComboObj();
+      break;
+
+    case MAIN_SOLVER::DISC_ADJ_FEM:
+      solver[FEA_SOL]->Postprocessing(geometry, config, numerics_container[ZONE_0][INST_0][MESH_0][FEA_SOL], true);
+
+      direct_output->SetHistoryOutput(geometry, solver, config, config->GetTimeIter(),
+                                      config->GetOuterIter(), config->GetInnerIter());
+      ObjFunc = solver[FEA_SOL]->GetTotal_ComboObj();
+      break;
+
+    default:
+      break;
+  }
+}
+
+void CDiscAdjSinglezoneDriver::UpdateStates() {
+  /*--- Update the flow and turbulent conservative state variables, preparing for other updates. ---*/
+  direct_iteration->Preprocess(direct_output, integration_container, geometry_container, solver_container,
+                               numerics_container, config_container, surface_movement, grid_movement, FFDBox, ZONE_0,
+                               INST_0);
+}
+
+void CDiscAdjSinglezoneDriver::UpdateResiduals() {
+  /*--- Update the Euler, Navier-Stokes or Reynolds-averaged Navier-Stokes (RANS) residuals and objective function (no
+   * system solve). ---*/
+  integration[FLOW_SOL]->ComputeResiduals(geometry_container, solver_container, numerics_container, config_container,
+                                          FLOW_SOL, ZONE_0, INST_0);
+}
+
+void CDiscAdjSinglezoneDriver::UpdateTractions() {
+  /*--- Update the surface tractions. ---*/
+  direct_iteration->Postprocess(direct_output, integration_container, geometry_container, solver_container,
+                                numerics_container, config_container, surface_movement, grid_movement, FFDBox, ZONE_0,
+                                INST_0);
+}
+
+void CDiscAdjSinglezoneDriver::UpdateJacobians() {
+  /*--- Compute the approximate Jacobian for preconditioning. ---*/
+  solver[FLOW_SOL]->PrepareImplicitIteration(geometry, solver, config);
+}
+
+void CDiscAdjSinglezoneDriver::ApplyPreconditioner(const CSysVector<Scalar>& u, CSysVector<Scalar>& v) {
+  /*--- Use an approximate diagonal preconditioning based on the transpose of the primal Jacobian. ---*/
+  (*PrimalPreconditioner)(u, v);
+
+  /*--- Apply a few FGMRES iterations in addition to the above preconditioner. ---*/
+  v.SetValZero();
+
+  Scalar KrylovPreEps = KrylovPreTol;
+  auto MaxIter = 5;
+  solver[FLOW_SOL]->System.FGMRES_LinSolver(u, v, *PrimalJacobian, *PrimalPreconditioner, KrylovPreTol, MaxIter,
```

Comment:
Some regression tests are failing due to a type conversion error from `CSysVector<double>` to 'CSysVector<float>'. Can someone please advise on the best way to fix this?

Here is the error:
```
[699/745] Compiling C++ object SU2_CFD/src/libSU2core.a.p/drivers_CDiscAdjSinglezoneDriver.cpp.o
FAILED: SU2_CFD/src/libSU2core.a.p/drivers_CDiscAdjSinglezoneDriver.cpp.o 
ccache c++ -ISU2_CFD/src/libSU2core.a.p -ISU2_CFD/src -I../SU2_CFD/src -I../externals/CLI11 -I../externals/cgns/adf -I../externals/cgns/adfh -Iexternals/cgns -I../externals/cgns -Iexternals/cgns/hdf5 -I../externals/cgns/hdf5 -I../externals/metis/include -I../externals/metis/GKlib -I../externals/metis/libmetis -I../externals/parmetis/include -I../externals/parmetis/libparmetis -I../externals/mel -ICommon/src -I../Common/src -I/usr/lib/x86_64-linux-gnu/openmpi/include -I/usr/lib/x86_64-linux-gnu/openmpi/include/openmpi -fdiagnostics-color=always -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -Wnon-virtual-dtor -Wextra -Wpedantic -Werror -std=c++11 -O3 -fPIC -fopenmp -pthread -Wno-empty-body -Wno-ignored-qualifiers -Wno-unused-parameter -Wno-deprecated-declarations -Wno-error=cast-function-type -Wno-cast-function-type -Wno-error=stringop-truncation -Wno-stringop-truncation -Wno-error=missing-field-initializers -Wno-missing-field-initializers -Wno-error=class-memaccess -Wno-class-memaccess -DHAVE_CGNS -DNDEBUG -DUSE_MIXED_PRECISION -DHAVE_MPI -DHAVE_METIS -DHAVE_PARMETIS -MD -MQ SU2_CFD/src/libSU2core.a.p/drivers_CDiscAdjSinglezoneDriver.cpp.o -MF SU2_CFD/src/libSU2core.a.p/drivers_CDiscAdjSinglezoneDriver.cpp.o.d -o SU2_CFD/src/libSU2core.a.p/drivers_CDiscAdjSinglezoneDriver.cpp.o -c ../SU2_CFD/src/drivers/CDiscAdjSinglezoneDriver.cpp
../SU2_CFD/src/drivers/CDiscAdjSinglezoneDriver.cpp: In member function ‘void CDiscAdjSinglezoneDriver::ApplyPreconditioner(const CSysVector<double>&, CSysVector<double>&)’:
../SU2_CFD/src/drivers/CDiscAdjSinglezoneDriver.cpp:827:45: error: cannot convert ‘const CSysVector<double>’ to ‘const VectorType&’ {aka ‘const CSysVector<float>&’}
  827 |   solver[FLOW_SOL]->System.FGMRES_LinSolver(u, v, *PrimalJacobian, *PrimalPreconditioner, KrylovPreTol, MaxIter,
      |                                             ^
      |                                             |
      |                                             const CSysVector<double>
In file included from ../SU2_CFD/src/drivers/../../include/drivers/../integration/../solvers/CSolver.hpp:52,
                 from ../SU2_CFD/src/drivers/../../include/drivers/../integration/CIntegration.hpp:34,
                 from ../SU2_CFD/src/drivers/../../include/drivers/CDriver.hpp:33,
                 from ../SU2_CFD/src/drivers/../../include/drivers/CSinglezoneDriver.hpp:30,
                 from ../SU2_CFD/src/drivers/../../include/drivers/CDiscAdjSinglezoneDriver.hpp:30,
                 from ../SU2_CFD/src/drivers/CDiscAdjSinglezoneDriver.cpp:28:
../SU2_CFD/src/drivers/../../include/drivers/../integration/../solvers/../../../Common/include/linear_algebra/CSysSolve.hpp:325:52: note:   initializing argument 1 of ‘long unsigned int CSysSolve<ScalarType>::FGMRES_LinSolver(const VectorType&, CSysSolve<ScalarType>::VectorType&, const ProductType&, const PrecondType&, ScalarType, long unsigned int, ScalarType&, bool, const CConfig*) const [with ScalarType = float; CSysSolve<ScalarType>::VectorType = CSysVector<float>; CSysSolve<ScalarType>::ProductType = CMatrixVectorProduct<float>; CSysSolve<ScalarType>::PrecondType = CPreconditioner<float>]’
  325 |   unsigned long FGMRES_LinSolver(const VectorType& b, VectorType& x, const ProductType& mat_vec,
      |   
```

Reply:
Fixed in commit [fe7f82c](https://github.com/su2code/SU2/pull/1750/commits/fe7f82cee2af2cd18f583f40316e09de4f701dd5)